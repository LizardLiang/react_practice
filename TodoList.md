#This is a Todo list to my page#

- Bugs:

  1. button style broke when too small
  2. animate won't work due to using fixed coords in code

- Features to add
  1. Favorite list for stock id
  2. Overall theme remake
