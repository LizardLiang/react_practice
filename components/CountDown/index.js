export * from './CountDown.jsx'

import {createStore} from 'redux' // import store component

const defaultDate = {
    title:  'New Year!!',
    year:   2021,
    month:  1,
    day:    1,
    hour:   0,
    minute: 0,
    second: 0,
    isAnnual: true
}

// create data
const Datas = {
    dates: [
    ],
    curUser: ''
}

// define actions
const addDate = article => ({type: 'addDate', payload: article})
const deleteDate = article => ({type: 'deleteDate', payload: article})
const setUser = article => ({type: 'setUser', payload: article})
const clearTimer = article => ({type: 'clearTimer', payload: article})

// define actual method for actions
const rootReducer = (state = Datas, action) => {
    switch(action.type){
        case 'addDate':
            return {...state, dates: [...state.dates, action.payload]}
        case 'deleteDate':
            return {...state, dates: state.dates.filter((item => item.title !== action.payload))}
        case 'setUser':
            return (Object.assign({}, state, {curUser: action.payload}))
        case 'clearTimer':
            return (Object.assign({}, state, {curUser: '', dates: [defaultDate]}))
        default:
            return state
    }
}

// establish store
const store = createStore(rootReducer)

export {store, addDate, deleteDate, defaultDate, setUser, clearTimer}

// export default Stores