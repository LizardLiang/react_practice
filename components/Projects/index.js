export * from "./test_program.jsx"
export * from "./web_game.jsx"
export * from "./StockChart.jsx"
export * from "./projects.jsx"