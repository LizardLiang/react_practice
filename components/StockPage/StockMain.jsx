import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import styles from 'styled-components'

const HomeLink = styles(Link) `
    float:          right;
`

const DivContainer = styles.div`
    position:       absolute;
    top:            0;
    left:           0;
    width:          100vw;
    height:         100vh;
    background:     rgb(32, 34, 37);
`

const InputDiv = styles.div `
    position:       relative;
    width:          868px;
    height:         auto;
    margin:         0 auto;
`

const Span = styles.span `
    color:          white;
    display:        block;
    text-align:     center
`

const Button = styles.button `
    background:     grey;
    color:          white;


    &:hover {
        background: white;
        color:      black;
    }
`

const Svg = styles.svg `
    @media screen and (min-width: 1080px){
        width:          50%;
        height:         auto;
        overflow:       visible;
        margin-bottom:  10%;
        margin-top:     10%;
        margin-left:    auto;
        margin-right:   auto;
        display:        ${props => props.svgDisplay || 
                                    "block"};
    }
    @media screen and (max-width: 600px){
        width:          50vw;
        height:         25vh;
        overflow:       visible;
        margin-bottom:  10%;
        margin-top:     10%;
        margin-left:    auto;
        margin-right:   auto;
        display:        ${props => props.svgDisplay || 
                                    "block"};
    }
`

const Text = styles.text `
    fill:               white;
    font-size:          20%;
`

const GridLine = styles.line `
    stroke:             grey;
    stroke-opacity:     .5;
`

// calculate axis x text position
function calxPosition(num, leng){
    return (num / leng) * 100
};


// add lines to grid
function add_y_line(){
    let list=[]
    for (let i=0; i < 4; ++i){
        <GridLine x1='0' x2='100' y1={i * 25} y2={i * 25}/>
    }
}

// add axis y text
function add_y_grid(min, max){
    let list = []
    for(let i=1; i < 5; ++i){
        list.push(
            <Text x='-5' y={100 - i * 25}>
                {(max - min) / 4 * i + min}
            </Text>
        )
    }
    return (
        list
    )
}

// dealing with line and data
class ChartLine extends Component{
    render(){
        return (
            <polyline
                fill="none"
                stroke="green"
                stroke-width="1"
                points={this.props.points} />
        )
    }
}

function add_x_text(labels){
    let list = labels.map((value, index)=>{
        if(index % 7 == 0 || index == (labels.length - 1)){
            return(
                <Text 
                    x={calxPosition(index, labels.length) - 10} 
                    y='105'>
                    {value}
                </Text>
            )
        }
    })
    return list
}

// dealing with svg
class SvgChart extends Component    {
    render(){
        let pointList = this.props.points.split(' ')
        pointList = pointList.slice(0, -1)
        const datacircle = 
            pointList.map((value, index)=>{
                const point = value.split(',')
                return(
                    <circle cx={calxPosition(index, pointList.length)}
                        cy={parseFloat(point[1])} r='1' fill='green'/>
                )
            })
        return (
            <Svg    
                id='stockChart'
                viewBox='0 0 100 100'
                preserveAspectRatio='none'
                svgDisplay={this.props.SvgDisplay}
                style={styles.MainContainer}>
                <g style={{stroke: "grey"}}>
                    <line 
                        x1='0'
                        x2='100'
                        y1='100'
                        y2='100'/>
                </g>
                <g style={{stroke: "grey"}}>
                    <line
                        x1='0'
                        x2='0'
                        y1='0'
                        y2='100'/>
                </g>
                {add_y_line()}
                {add_y_grid(this.props.min, this.props.max)}
                {add_x_text(this.props.xlabel)}
                {datacircle}
                <ChartLine points={this.props.points} />
            </Svg>
        )
    }
}

class StockMain extends Component{
    constructor(props){
        super(props)
        this.state = {
            stockID:        '',
            minNum:         0,
            maxNum:         0,
            x_texts:        [],
            points:         '',
            reverseHeight:  0,
            svgDisplay:     'none'
        }
    }

    calyPosition(num){
        return ((num - this.state.minNum) / (this.state.maxNum - this.state.minNum)) * 100
    }

    FetchStockInfo = () => {
        let id = this.state.stockID
        if (id == '' || isNaN(id) == true){
            alert('Stock ID must be numbers')
        }
        else{
            let url = 
            'http://114.32.157.74/PythonFlask/api/test?id='
            fetch(url + id)
            .then((response)=>{
                return response.json()
            }).then((item)=>{
                let values = item['value']
                this.setState({x_texts: item['date']})
                let point = ''
                values = values.map((value)=>{
                    return parseFloat(value)
                })
                this.setState({maxNum: Math.max(...values), minNum: Math.min(...values)})
                for (let i=0; i < values.length; i++){
                    let calx = calxPosition(i, values.length)
                    point += calx
                    point += ','
                    let calNum = this.calyPosition(values[i])
                    point += (100 - calNum) // 100 - x to fit coordinate
                    point += ' '
                }
                this.setState({points: point, svgDisplay: 'block'})
            }).catch((error)=>{
                this.setState({svgDisplay: 'none'})
            })
        }
    }

    handleInputEvent = () => (event) => {
        this.setState({stockID: event.target.value})
    }

    render(){
        const Styles = {
            headerSpanContainer:{
                display:    'block',
                textAlign:  'center',
                color:      'white'
            },
            centerDiv:{
                width:  '50%',
                margin: '2% auto'
            }
        }
        return(
            <DivContainer>
                <Span>輸入股票代碼以取得股票資訊</Span>
                <HomeLink to='/'>
                    <Button>
                        回首頁
                    </Button>
                </HomeLink>
                <InputDiv>
                    <input 
                        aria-label='input'
                        type='text'
                        name='stockID'
                        value={this.state.stockID}
                        onChange={this.handleInputEvent()}>
                    </input>
                    <Button 
                        onClick={this.FetchStockInfo}
                        arial-label='searchButton'>
                            搜尋
                    </Button>
                </InputDiv>
                <SvgChart 
                    max={this.state.maxNum}
                    min={this.state.minNum}
                    SvgDisplay={this.state.svgDisplay}
                    points={this.state.points}
                    xlabel={this.state.x_texts}/>
                
            </DivContainer>
        )
    }
}

export { StockMain }