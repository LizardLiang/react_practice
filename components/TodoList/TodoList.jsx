import React, {useState, useEffect} from 'react'
import {Provider, useSelector, useDispatch} from 'react-redux'
import styled, {keyframes} from 'styled-components'
import {store} from './'

const EditMainDiv = styled.div ` 
    width: 70%;
    max-width: 1344px;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    border: solid 1px white;
    border-radius: 5px;
    background:rgb(47, 49, 54);

    div {
        padding: 5px;
        border-radius: 5px;
    }

    input {
        color: black;

        @media screen and (max-width: 600px){
            font-size: 1rem;
        }
    }
`

const EditHeadDiv = styled.div ` 
    display: flex;
    font-size: 2rem;
`

const EditTitleDiv = styled.div ` 
    width: 100%;
    display: flex;
    background-color: rgb(64, 68, 75);
    font-size: 1.5rem;
`

const EditTitleInput = styled.input ` 
    width: 50%;
    border-radius: 5px;
    margin-left: 5px;
`

const EditDateDiv = styled.div ` 
    width: 100%;
    display: flex;
    justify-content: flex-start;
    background-color: rgb(64, 68, 75);
    font-size: 1.5rem;
    
    .confirmB {
        margin-left: auto;
    }
`

const EditDateInput = styled.input `
    border-radius: 5px;
    margin-left: 5px;
`

const EditButton = styled.button ` 
    margin-top: 0;
    margin-bottom: 0;
    margin-right: 5px;

    color: white;
    background-color: rgb(64, 68, 75);
    border: solid grey 1px;
    border-radius: 5px;

    @media screen and (max-width: 600px){
        font-size: 1rem;
    }
`

const EditBlock = (props) => {
    const [title, setTitle] = useState("")
    const [date, setDate] = useState("")
    return (
        <EditMainDiv>
            <EditHeadDiv>新增代辦事項</EditHeadDiv>
            <EditTitleDiv>
                標題:
                <EditTitleInput 
                    id="Title" placeholder={`輸入標題`} 
                    onChange={ e => setTitle(e.target.value)} />
            </EditTitleDiv>
            <EditDateDiv>
                日期:
                <EditDateInput 
                    type="Date" id="Date"
                    onChange={e => setDate(e.target.value)} />
                {/* <EditDateInput placeholder={`輸入月`}/>
                <EditDateInput placeholder={`輸入日`}/> */}
                <EditButton className="confirmB" onClick={() => {props.addMsg(title, date)}}>
                    確認
                </EditButton>
                <EditButton>取消</EditButton>
            </EditDateDiv>
        </EditMainDiv>
    )
}

const ContMain = styled.div `
    display: flex;
    width: 100%;
    font-size: 1.5rem;
    padding: 5px 5px 5px 5%;
    background-color: rgb(64, 68, 75);
    border-radius: 5px;

    .div {
        padding: 0;
    }

    .Date {
        margin-left: auto;
    }
`

const ContNameDiv = styled.div `
    display: flex;
    .SO {
        margin-right: 12px;
    }
`

const ContDateDiv = styled.div `
    display: flex;
    align-items: flex-end;

    @media screen and (max-width: 600px){
        font-size: 1rem;
    }

    .SO {
        margin-right: 12px;
    }
`

const ContSetting = styled.div ` 
    display: inline-block;
    justify-content: center;
    align-items: center;
    text-align: center;
    position: relative;
    white-space: nowrap;
`

const FadeIn = keyframes `
    0%{
        opacity: 0;
        width: 0px;
    }

    100% {
        opacity: 1;
        width: 160px;
    }
`

const ContButton = styled.button ` 
    background: none;
    color: white;
    display: flex;
    align-items: flex-end;

    @media screen and (max-width: 600px){
        font-size: 1rem;
    }

    & :focus{
        outline: none;
        div {
            display: flex;
            visibility: visible;
            opacity: 1;
            width: 160px;
            animation: ${FadeIn} 1s;
        }
    }
` 

const ContSettings = styled.div ` 
    position: absolute;
    flex-direction: column;
    width: 0px;
    z-index: 1;
    background: rgb(24, 25, 28);
    border-radius: 5px;
    display: none;
    visibility: hidden;
    opacity: 0;
    /* transition: opacity 500ms; */

    &:hover {
        display: flex;
        opacity: 1;
        visibility: visible;
        width: 160px;
    }
`

const ContRadioBlk = styled.div `
    display: flex;
    align-items: center;
    padding: 5px;
    margin: 5px;
    justify-content: space-evenly;
    border-radius: 5px;

    &:hover{
        background: rgb(92, 111, 177);
    }

    button {
        height: 100%;
    }
`

const TodoContent = (props) => {
    const [display, setDis] = useState('none')

    return (
        <ContMain>
            <ContNameDiv>
                <span className="SO">{props.so}</span>
                <span>{props.name}</span>
            </ContNameDiv>
            <ContDateDiv className="Date">
                <span className="SO">期限:</span>
                <span>{props.Date}</span>
            </ContDateDiv>
            <ContSetting>
                <ContButton 
                    onClick={()=>{display == 'none' ? setDis('flex') : setDis('none')}}>
                    more
                    <ContSettings>
                        <ContRadioBlk
                            onClick={()=>{props.setSts(props.name, props.status, 0)}}>
                            <input
                                type="radio" 
                                name={`status${props.so}`} 
                                checked={props.status == 0 ? true : false}/>
                                進行中
                        </ContRadioBlk>
                        <ContRadioBlk
                            onClick={()=>{props.setSts(props.name, props.status, 1)}}>
                            <input
                                type="radio" 
                                name={`status${props.so}`}
                                checked={props.status == 1 ? true : false}/>已完成
                        </ContRadioBlk>
                        <ContRadioBlk
                            onClick={()=>{props.setSts(props.name, props.status, 2)}}>
                            <input
                                type="radio" 
                                name={`status${props.so}`}
                                checked={props.status == 2 ? true : false}/>暫停中
                        </ContRadioBlk>
                    </ContSettings>
                </ContButton>
            </ContSetting>
        </ContMain>
    )
}

const MainDiv = styled.div `
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    padding-top: 5rem;
    color: white;
    min-height: 100vh;
`

const CateBlock = styled.div `
    display: flex;
    flex-direction: column;
    width: 70%;
    max-width: 1344px;
    height: auto;
    min-height: 10%;
    border: white solid 1px;
    border-radius: 5px;
    margin: 5px;
    background:rgb(47, 49, 54);
`

const CateTitle = styled.div `
    width: 100%;
    height: auto;
    font-size: 2rem;
    padding-left: 5px;
`

const NoItemSpan = styled.span`
    display: flex;
    width: 100%;
    font-size: 1.5rem;
    padding: 5px 5px 5px 5%;
    background-color: rgb(64, 68, 75);
    border-radius: 5px;
`

function TodoList() {
    // Three different task status
    const [actives, setActive] = useState([])
    const [finishes, setFinished] = useState([])
    const [inactives, setInactive] = useState([])

    // Use selector to get store data
    const ActiveMsgs = useSelector(state => state.ActiveMsg)
    const FinishMsgs = useSelector(state => state.FinishedMsg)
    const InactiMsgs = useSelector(state => state.InactiveMsg)

    // Set dispatch to modify store data
    const MsgDispatch = useDispatch()

    let arrs = [actives, finishes, inactives];
    let arr_funcs = [setActive, setFinished, setInactive];

    useEffect(()=>{
        setActive(ActiveMsgs)
        setFinished(FinishMsgs)
        setInactive(InactiMsgs)
    }, [])


    let changeStatus = (name, old, newSts) => {
        // get specify target
        let target = arrs[old].filter(item=>item.Name===name)[0]

        // remove target from old list
        let func = arr_funcs[old]
        func(arrs[old].filter(item=>item.Name!==name))

        // append target to new list with new status
        target.status = newSts
        let newfunc = arr_funcs[newSts]
        newfunc(prev => [...prev, target])
    }

    const addMsg = (title, date) => {
        setActive(old => [...old, {Name: title, Date: date, status: 0}])
        MsgDispatch({type: 'addMsg', payload: {Name: title, Date: date, status: 0}})
    }

    return(
        <MainDiv>
            <CateBlock>
                <CateTitle>Active</CateTitle>
                {/* Get active list */}
                {actives.length > 0 ? actives.map(((value, index)=>{
                    return <TodoContent name={value.Name} Date={value.Date} so={index + 1} status={value.status} setSts={changeStatus} />})) : 
                    <NoItemSpan>沒有項目</NoItemSpan>
                }
            </CateBlock>
            <CateBlock>
                <CateTitle>Finished</CateTitle>
                {/* Get finshed list */}
                {finishes.length > 0 ? 
                    finishes.map(((value, index)=>{
                    return <TodoContent name={value.Name} Date={value.Date} so={index + 1} status={value.status} setSts={changeStatus} />})) : 
                    <NoItemSpan>沒有項目</NoItemSpan>
                }
            </CateBlock>
            <CateBlock>
                <CateTitle>Inactive</CateTitle>
                {/* Get inactive list */}
                {inactives.length > 0 ? 
                    inactives.map(((value, index)=>{
                    return <TodoContent name={value.Name} Date={value.Date} so={index + 1} status={value.status} setSts={changeStatus} />})) : 
                    <NoItemSpan>沒有項目</NoItemSpan>
                }
            </CateBlock>
            <EditBlock addMsg={addMsg} />
        </MainDiv>
    )
}

const FinalList = () => {
    return(
        <Provider store={store}>
            <TodoList />
        </Provider>
    )
} 

export default FinalList