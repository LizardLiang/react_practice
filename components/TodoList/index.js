export * from './TodoList.jsx'

import {createStore} from "redux"

// Seperate these to three different part
const Messages = {
    ActiveMsg: [{Name: 'Active', Date: '2020-01-01', status: 0}],
    FinishedMsg: [{Name: 'Finished', Date: '2020-01-01', status: 1}],
    InactiveMsg: [{Name: 'Inactive', Date: '2020-01-01', status: 2}]
}

const addMessage = article => ({type: 'addMsg', payload: article})
const delMessgae = article => ({type: 'delMsg', payload: article})

const rootReducer = (state = Messages, action) => {
    switch(action.type){
        case 'addMsg': 
            return {...state, ActiveMsg: [...state.ActiveMsg, action.payload]}
        default:
            return state
    }
}

const store = createStore(rootReducer)

window.store = store;
window.addMessage = addMessage;

export {store, addMessage, delMessgae}