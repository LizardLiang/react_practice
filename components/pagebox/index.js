export * from "./info.jsx"
export * from "./experience.jsx"
export * from "./project.jsx"