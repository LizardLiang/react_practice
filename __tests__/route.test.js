import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import { StockMain } from '../components/StockPage'
import { Simulate } from 'react-dom/test-utils'

import {TodoList} from '../components/TodoList/TodoList.jsx'
 
describe('Stock Page', () => {
  test('snapshot renders', () => {
    const inputContainer = render(<StockMain/>)
    const inst = inputContainer.getByLabelText('input')
    const sButton = searchButton.getByLabelText('searchButton')

    fireEvent.change(inst, {target: {value: '123'}})
    fireEvent.click(sButton)
    expect(inst.value).toBe('123')
  })
})