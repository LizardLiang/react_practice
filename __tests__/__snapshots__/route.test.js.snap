// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter snapshot renders 1`] = `
<div>
  123
</div>
`;
